adr/adr-1.md

```diff
@@ -0,0 +1,124 @@
+
+<!-- This is an optional element. Feel free to remove. -->
+### Consequences
+
+* 👍 Nex node no longer needs compiled-in dependencies used by individual workload types
+* 👍 Multitenant agents mean there is significantly less per-workload overhead, especially on function types
+* 👍 Moving a large amount of code to the agents can dramatically cut down on the complexity of the core Nex node codebase
+* Makes `preflight` a little bit more complicated as it now needs to download agents for first-party workload types
+* Creates breaking changes pretty much everywhere, making it so at least the first two (JavaScript, Native) agents would have to be built along with the Nex refactor.
+
+## Detailed Design
+In this new design, the `nex` binary is distributed on its own. During the `preflight` command we can download all of the relevant agents the same way we download the `firecracker` binary and the CNI plugin files.
+
+**All agents are multi-tenant**
+
+The Nex node configuration file will change slightly to accommodate this. Inside the `workload_types` configuration object, instead of having just a list of the name of each supported workload type, we will now have configuration per type:
+
+```json
+{
+    ...
+    "workload_types": [
+        {
+            "name": "native",
+            "agent_path": "/opt/agents/native",
+            "argv": [...],
+            "env": {
+                "a": "b"
+            }                     
+        },
+        {
+            "name": "javascript",
+            "agent_path": "/opt/agents/javascript"           
+        }    
+    ] 
+}
+```
+Note that if an `agent_path` is not supplied for the supported workload type, then Nex will look for the agent at `./agents/{name}`, relative to the Nex startup directory. Running `preflight` against a configuration file with first-party workload types defined can automatically download those agents and put them in the appropriate location.
+
+In some cases, an agent may support artifacts that are either regular files or **archives**. When this occurs (e.g. in the case of the `JavaScript` agent, which can run a single `.js` file or the contents of a tarball), it is up to the agent to detect whether the artifact is an archive or not and to deal with it properly.
+
+Poorly formed or invalid archives should be rejected by the agent in response to the start request.
+
+All agents are spawned by the `Nex` node process at startup and must _not_ be started manually outside the Nex node process (e.g. they are _not_ sidecars). Each agent will be supplied with the following environment variables which allow it to connect to the Nex internal NATS "bus":
+
+* `INT_NATS_HOST` - Host (IP) of the Nex internal NATS server. This will almost always be `0.0.0.0` or `127.0.0.1`, and is determined by the node configuration file.
```

Comment:
Slight nit, but I suggest adding `NEX_` as a prefix (to contextualize the target) and since this is from the agent's perspective, could it be just `NEX_NATS_HOST`? Meaning does the `INT` (for internal) add any meaning to this?